Build a wallet using js - JS Bounty (#1752)
XRPLF/xrpl-dev-portal: content/tutorials/build-apps/build-a-browser-wallet-in-js.md

![Home Page Screenshot](img/js-wallet-home.png)

This application can:

- Show updates to the XRP Ledger in real-time.
- View any XRP Ledger account's activity, including showing how much XRP was delivered by each transaction.
- Show how much XRP is set aside for the account's [reserve requirement](reserves.html). 
- Send [direct XRP payments](direct-xrp-payments.html), and provide feedback about the intended destination address, including:
    - Displaying your account's available balance
    - Verifying that the destination address is valid
    - Validating the account has enough XRP to send
    - Allowing you to specify a destination tag

## Steps

Before you begin, make sure you have the prerequisites installed. Check your node version by running `node -v`. If necessary, [download Node.js](https://nodejs.org/en/download/).

**Tip:** If you get stuck while doing this tutorial, or working on another project, feel free to ask for help in the XRPL's [Developer Discord](https://discord.com/invite/KTNmhJDXqa).

### 1. Set up the project

1. Navigate to the directory that you want to create the project in.

> 2. Run 
>    ```sh
>     yarn create vite-app <project-name>
>     ```
>     to create a new project.

Reviewer: This is still broken. It doesn't display as a code block, and the `sh` shows as if it's part of the command.
![(Screenshot showing the problem)](https://user-images.githubusercontent.com/7515597/232926723-2e8f6351-0a38-4bbb-9174-c9a9194e505d.png)

```suggestion
2. Create a new Vite project:

        yarn create vite-app simple-xrpl-wallet

```

3. Create or modify the file `package.json` to have the following contents:

{{ include_code("_code-samples/build-a-wallet/js/package.json", language="js") }}

   - Alternatively you can also do `yarn add <package-name>` for each individual package to add them to your `package.json` file.
  

> 4. Run `yarn` to install the packages.

Reviewer: I accidentally glossed over this step while following the tutorial. To make it harder to miss, let's put the command in a code block:

```suggestion
4. Install dependencies:

        yarn
```

5. Create a new file `.env` in the root directory of the project and add the following variables:

```bash
CLIENT="wss://s.altnet.rippletest.net/" // or any other server
EXPLORER_NETWORK="testnet" // or "mainnet"
SEED="s████████████████████████████"  // Replace with your seed
```

6. Change the seed to your own seed. You can get credentials from [the Testnet faucet](xrp-test-net-faucet.html).

7. Set up a Vite bundler. Create a file named `vite.config.js` in the root directory of the project and fill it with the following code:

{{ include_code("_code-samples/build-a-wallet/js/vite.config.js", language="js") }}

This example includes the necessary configuration to make [xrpl.js work with Vite](https://github.com/XRPLF/xrpl.js/blob/main/UNIQUE_SETUPS.md#using-xrpljs-with-vite-react).

8. Add script to `package.json`

> In your `package.json` file, add:

Reviewer: If you pasted the file contents from step 3, this is already there. Probably best to clarify this so people don't get confused and try to add it a second time.

 ```suggestion
In your `package.json` file, add the following section if it's not there already:
```

    - Verifying that the destination address is valid
    - Validating the account has enough XRP to send
    - Allowing you to specify a destination tag

## Steps

Before you begin, make sure you have the prerequisites installed. Check your node version by running `node -v`. If necessary, [download Node.js](https://nodejs.org/en/download/).

**Tip:** If you get stuck while doing this tutorial, or working on another project, feel free to ask for help in the XRPL's [Developer Discord](https://discord.com/invite/KTNmhJDXqa).

### 1. Set up the project

1. Navigate to the directory that you want to create the project in.
2. Create a new Vite project:

```bash
yarn create vite simple-xrpl-wallet --template vanilla
```

3. Create or modify the file `package.json` to have the following contents:

{{ include_code("_code-samples/build-a-wallet/js/package.json", language="js") }}

   - Alternatively you can also do `yarn add <package-name>` for each individual package to add them to your `package.json` file.
  
4. Install dependencies:

```bash
yarn
```

5. Create a new file `.env` in the root directory of the project and add the following variables:

```bash

> CLIENT="wss://s.altnet.rippletest.net/" // or any other server
> EXPLORER_NETWORK="testnet" // or "mainnet"
> SEED="s████████████████████████████"  // Replace with your seed

Reviewer 2: The .env file doesn't actually support comments, so this causes an error for me when the `// or any other server` is left in. Also it needs the websocket extension 51233. I'd recommend having it just work out of the box, except for the seed which you have a step right after to fix :)
```suggestion
CLIENT="wss://s.altnet.rippletest.net:51233/"
EXPLORER_NETWORK="testnet"
SEED="s████████████████████████████"
```

Reviewer 3: It should support comments starting with `#`.

Author: I have removed the comments

```

6. Change the seed to your own seed. You can get credentials from [the Testnet faucet](xrp-test-net-faucet.html).

7. Set up a Vite bundler. Create a file named `vite.config.js` in the root directory of the project and fill it with the following code:

{{ include_code("_code-samples/build-a-wallet/js/vite.config.js", language="js") }}

This example includes the necessary configuration to make [xrpl.js work with Vite](https://github.com/XRPLF/xrpl.js/blob/main/UNIQUE_SETUPS.md#using-xrpljs-with-vite-react).

8. Add script to `package.json`

In your `package.json` file, add the following section if it's not there already:

```json
"scripts": {
    "dev": "vite"
}
```

### 2. Create the Home Page (Displaying Account & Ledger Details)

In this step, we create a home page that displays account and ledger details.

![Home Page Screenshot](img/js-wallet-home.png)

1. Make a new folder named `src` in the root directory of the project.

> 2. If not already present, create new files in the root folder named `index.html`, `index.js` and `index.css`.

Reviewer 2: I'd recommend swapping the order of these two steps because when running through these steps I got confused and put these 3 files into the `src` folder, which doesn't work. If we have them create the files first, then they won't accidentally add them as a sub-folder file.

Author: done

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/get-wallet-details.js", language="js") }}

6. Now, let's add the code to `index.js` file to fetch the account and ledger details and display them on the home page. Copy the code written below to the `index.js` file. Here we render the wallet details using the function we defined in `get-wallet-details.js`. In order to make sure we have up to date ledger data, we are using the [ledger stream](subscribe.html#ledger-stream) to listen for ledger close events.

{{ include_code("_code-samples/build-a-wallet/js/index.js", language="js") }}

The one other thing we do here is add events to two buttons - one to send XRP and one to view transaction history. They won't work just yet—we'll implement them in the next steps.

Now the application is ready to run. You can start it in dev mode using the following command:

```bash
yarn dev

> ```

Reviewer 2: At this point the code crashes because `index.js` refers to `./src/helpers/render-xrpl-logo` but we haven't defined that yet.
<img width="1408" alt="image" src="https://user-images.githubusercontent.com/19694186/235007399-4155e403-e270-42e2-a2e9-bc011be09417.png">

Reviewer 2: After adding BOTH `render-xrpl-logo.js` and `xrpl.svg` it renders properly.

Author: I've added this: 

Also, add the following two files, one in the `helpers` folder [render-xrpl-logo.js]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/src/helpers/render-xrpl-logo.js) and for the other one create a new folder named `assets` in `src` directory and add [xrpl.svg]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/src/assets/xrpl.svg) file. These files are used to render the XRPL logo for aesthetic purposes.

Reviewer 2: Works up to here following steps as written now :)

4. Now that we have the HTML code, let's add the JavaScript code. In the `helpers` folder, create a new file named `submit-transaction.js` and copy the code written below to the file. In this file, we are using the [submit](submit.html) method to submit the transaction to the XRPL. Before submitting every transaction needs to be signed by a wallet, learn more about [signing](sign.html) a transaction.

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/submit-transaction.js", language="js") }}

5. Now back to the `send-xrp.js` file, copy the code written below to the file. In this piece of code we are first getting all the DOM elements from HTML and adding event listners to update & validate the fields based on the user input. Using `renderAvailableBalance` method we display the current available balance of the wallet. `validateAddress` function validates the user address, and the amount is validated using a regular expression. When all the fields are filled with correct inputs, we call the `submitTransaction` function to submit the transaction to the ledger.

{{ include_code("_code-samples/build-a-wallet/js/src/send-xrp/send-xrp.js", language="js") }}

> 

Reviewer 2: I'd recommend a small section here showcasing what the app should now be able to do. E.g. A screenshot, and something like "You can now click 'Send XRP' to try creating your own transaction! You can use this example to send XRP to the testnet faucet to try it out"

Testnet faucet account: rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe
Amount: 100 
Destination Tag: (Not usually necessary unless you're paying an account tied to an exchange)

Author: Added

---
parent: build-apps.html
targets:
  - en
  - ja # TODO: translate this page
blurb: Build a graphical browser wallet for the XRPL using Javascript.
---
# Build A Browser Wallet Using JS
<!-- STYLE_OVERRIDE: wallet -->

This tutorial demonstrates how to build a browser wallet for the XRP Ledger using the Javascript programming language and various libraries. This application can be used as a starting point for building a more complete and powerful application, as a reference point for building comparable apps, or as a learning experience to better understand how to integrate XRP Ledger functionality into a larger project.

## Prerequisites

To complete this tutorial, you should meet the following guidelines:

> 1. You have [Node.js](https://nodejs.org/en/download/) v10.15.3 or higher installed.

Reviewer: I suggest a minimum version of Node.js v14 because xrpl.js isn't regularly tested on versions older than that. Maybe even add a recommended version of v16 LTS or something.

Author: Added

> 2. You have [Yarn](https://yarnpkg.com/en/docs/install) (v1.17.3 or higher) or [NPM](https://www.npmjs.com/get-npm) (v6.4.1 or higher) installed.

Reviewer: It's nice that the tutorial supports both, but I think it adds unnecessary complexity to have two examples every time one of these come up. I suggest picking one (doesn't matter to me which) and using that as the default for the examples. As long as you leave a note here about which one you use, I think that works—users who want to follow exact instructions can do that, and users who want to use the package manager of their choice can probably figure out how to adapt the `npm` examples to use `yarn` or vice-versa.

```suggestion
2. You have [NPM](https://www.npmjs.com/get-npm) (v6.4.1 or higher) installed. Alternatively, you can use [Yarn](https://yarnpkg.com/en/docs/install) (v1.17.3 or higher), but this tutorial's examples use NPM.
```

Author: Using Yarn and removing NPM :)

> 3. You are somewhat familiar with coding with JavaScript and have completed the [Get Started Using JavaScript](get-started-using-python.html) tutorial.

Reviewer: Copy-paste error here, linking the wrong tutorial:

```suggestion
3. You are somewhat familiar with coding with JavaScript and have completed the [Get Started Using JavaScript](get-started-using-javascript.html) tutorial.
```

## Source Code

You can find the complete source code for all of this tutorial's examples in the [code samples section of this website's repository]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/).

## Goals

At the end of this tutorial, you should be able to build a simple XRP wallet displayed below.

![Home Page Screenshot](img/js-wallet-home.png)

This application can:

- Show updates to the XRP Ledger in real-time.
- View any XRP Ledger account's activity, including showing how much XRP was delivered by each transaction.
- Show how much XRP is set aside for the account's [reserve requirement](reserves.html). 
- Send [direct XRP payments](direct-xrp-payments.html), and provide feedback about the intended destination address, including:
    - Displaying your account's available balance
    - Verifying that the destination address is valid
    - Validating the account has enough XRP to send
    - Allowing you to specify a destination tag

## Steps

> Before you begin, make sure you have the pre-requisites installed. Check your node version by running `node -v`. If you don't have node installed, you can download it [here](https://nodejs.org/en/download/).

Reviewer: Instead of using "here" as link text, I suggest rephrasing:

```suggestion
Before you begin, make sure you have the prerequisites installed. Check your node version by running `node -v`. If necessary, [download Node.js](https://nodejs.org/en/download/).
```

> ### Step 1: Setting up the project

Reviewer: Per our style guide, steps should be imperative sentences. (yes, the Python tutorial breaks this rule in some places. it should be fixed.) The word "Step" isn't necessary here either.

```suggestion
### 1. Set up the project
```

1. Navigate to the directory that you want to create the project in.
2. Run 
   ```sh
    yarn create vite-app <project-name>
    ```
    or
    ```sh
    npm init vite-app <project-name>
    ```

>     to create a new project.

Reviewer: This syntax doesn't work as intended. (You need an empty line to make it not treated as a hard wrap, and indented code fences are sketchy across Markdown parsers in general.) Right now it comes out like this:
![2023-03-31_134828_294540040](https://user-images.githubusercontent.com/7515597/229227584-6807fbdb-2357-4d2d-97a0-7b0bfa35e3ec.png)

I suggest rewording it to make the step text a description of what you're doing and use an indented code block. Also, I think it's better to provide an obvious placeholder name that will work if copy-pasted, rather than putting `<project-name>`, because the angle brackets may be interpreted as shell syntax. Finally, I think it's useful to link the Vite homepage to clarify what package you're using using to get started. In total, here's how I think it should look:

```suggestion
2. Create a new [Vite app](https://vitejs.dev/guide/) as follows:

        npm init vite-app simple-xrpl-wallet
```

(I suggest matching the name to the default name in the provided `package.json`)

Also, something that kind of confuses me is that `npm init vite-app simple-xrpl-wallet` and `yarn create vite-app` seem to do pretty different things. The `npm` one creates a nested `simple-xrpl-wallet` folder with `src/index.js` and stuff; the `yarn` one sets things up in the current working directory with `src/main.js` (not index.js). This has implications for some of the later steps of the project so it's even more important to pick one and be consistent with it.

> 3. Navigate to your project directory and copy this [package.json](_code-samples/build-a-wallet/js/package.json) file to your project.

Reviewer: That package.json link won't work. You _could_ link it like this, which would work after the PR is merged to the master branch (or if the appropriate fork/branch are passed in when building): 

```
copy this [`package.json`]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/package.json) file to your project.
```

But I suggest instead having the contents of the file in a code block the tutorial so users can just copy-paste, similar to with steps 5 and 7-8:

```suggestion
3. Create or modify the file `package.json` to have the following contents:

{{ include_code("_code-samples/build-a-wallet/js/package.json", language="js") }}
```

>         - Alternatively you can also do `yarn add <package-name>` or `npm install <package-name>` for each individual package to add them to your `package.json` file.

Reviewer: Our style is to use curly braces for variable segments of a code sample/command/etc (use [RFC6570](https://datatracker.ietf.org/doc/html/rfc6570) syntax if unclear).

Also, I suggest moving the bit about moving the `vite` script up to here while you're editing package.json, where it makes more sense (and it's clearer that it's already done if you copy whole the `package.json` over)

```suggestion
        - Alternatively you can also run `npm install {package-name}` for each individual package to add them to your `package.json` file. If you do, you should also add the following to the same file:
        
                "scripts": {
                    "dev": "vite"
                }

```

  
4. Run `yarn` or `npm i` to install the packages.

5. Create a new file `.env` in the root directory of the project and add the following variables:

```bash
CLIENT="wss://s.altnet.rippletest.net/" // or any other server
EXPLORER_NETWORK="testnet" // or "mainnet"
SEED="sEd7B2Jm7RBhhh5i88SH38aZB8ot8Kq" // Replace with your seed

> ```

Reviewer: To make sure the code block is properly indented within the numbered list, use blank lines before and after and 8 spaces rather than a code fence. Also, for placeholder seeds, I like to use a "censored" version with blocks rather than filling up the docs with examples of real seeds that you shouldn't use because they're publicly known:

```suggestion
        CLIENT="wss://s.altnet.rippletest.net/" // or any other server
        EXPLORER_NETWORK="testnet" // or "mainnet"
        SEED="s████████████████████████████" // Replace with your seed

```

> 6. Change the seed to your own seed. You can get a testnet account from [here](https://xrpl.org/xrp-test-net-faucet.html).

Reviewer: 1. Use use more informative link text
2. Use a relative link so it doesn't show as "external"

```suggestion
6. Change the seed to your own seed. You can get credentials from [the Testnet faucet](xrp-test-net-faucet.html).
```

7. Create a new file named `vite.config.js` if not present already in the root directory of the project.

> 8. Set up vite bundler, copy the code written below to the `vite.config.js` file. 

Reviewer: I suggest combining these steps:

```suggestion
7. Set up a Vite bundler. Create a file named `vite.config.js` in the root directory of the project and fill it with the following code:
```

{{ include_code("_code-samples/build-a-wallet/js/vite.config.js", language="js") }}

> xrpl.js requires some configuration to run with the vite bundler. You can find the configuration in [xrpl.js documentation](https://github.com/XRPLF/xrpl.js/blob/main/UNIQUE_SETUPS.md#using-xrpljs-with-vite-react).

Reviewer: This is confusing to me because it _looks_ like these steps have already been handled if you've followed the tutorial so far, namely setting up the `vite.config.js` and `package.json` files as instructed. Maybe rephrase it to clarify that this is what you've just done, not a separate thing you have to also do?

```suggestion
This example includes the necessary configuration to make [xrpl.js work with Vite](https://github.com/XRPLF/xrpl.js/blob/main/UNIQUE_SETUPS.md#using-xrpljs-with-vite-react).
```

1. Add script to `package.json`

In your `package.json` file, add:

```json
"scripts": {
    "dev": "vite"
}

> ```

Reviewer: I think you can remove this step (moving part of it under step 3 about adding the script if you're not copying `package.json` over exactly, as commented above)

> If you get stuck while doing this tutorial, or working on another project, feel free to ask for help in the XRPL's [Developer Discord](https://discord.com/invite/KTNmhJDXqa).

Reviewer: This feels like a random place for this comment. Maybe move it to the very start of "Steps"? Also, this would be a good opportunity to use a `**Tip:**` callout.

Author: Added

> ### Step 2: Creating the Home Page (Displaying Account & Ledger Details)

Reviewer: This header can also be shortened.

```suggestion
### 2. Create the home page
```

> In this step, we will create the home page that will display the account details and the ledger details.

Reviewer: ```suggestion
In this step, we create a home page that displays account and ledger details.
```

![Home Page Screenshot](img/js-wallet-home.png)

1. Make a new folder named `src` in the root directory of the project.

> 2. Create new files named `index.html`, `index.js` and `index.css`. 

Reviewer: When I ran it, the "create vite app" step already created:

- an `index.html` file at the top of the project
- a `src` folder containing `index.js` and `index.css`

I suspect that the intent is for the user to modify these files, rather than create new ones from scratch. So, these steps, telling the user to add new files, can be removed

Author: In my case it doesn't create index.js or index.css since I used yarn, but I think rephrasing it to `Create new files if not present already` will be a better.

> 3. Copy the contents of the [index.html](_code-samples/build-a-wallet/js/index.html), [index.js](_code-samples/build-a-wallet/js/index.js) and [index.css](_code-samples/build-a-wallet/js/index.css) files to your files.

Reviewer: These links won't work, either. I suggest splitting this into multiple steps (one for each file) that each use `{{ include_code() }}` to show the contents so users can copy-paste.

However, for `index.js`, step 6 does that already/again, so you can skip it here.

> Now that you've got the basic setup, let's discuss what we're going to do. For the home page, we want to:

Reviewer: ```suggestion
This basic setup creates a homepage and applies some visual styles. The goal is for the homepage to:
```

- Display our account info
- Show what's happening on the ledger
- And add a little logo for fun

> In order to make that happen, we need to look up our account state on the ledger. The next couple steps will walk you through that process.

Reviewer: ```suggestion
To make that happen, we need to connect to the XRP Ledger and look up the account and the latest validated ledger.
```

> 4. Make a new folder named helpers in the src directory, create a new file named `get-wallet-details.js` and copy the code written below to the file. In the function `getWalletDetails`, we are using the [account_info](https://xrpl.org/account_info.html) method to fetch the account details. We are also using the [server_info](https://xrpl.org/server_info.html) method to calculate the reserve base and increment. You can read more about the reserve requirement [here](https://xrpl.org/reserves.html).

Reviewer: - Use relative links here too
- For consistency with the rest of the site, "method" should be part of the link text.
- I suggested some other edits for brevity

```suggestion
4. In the `src/` directory, make a new folder named `helpers`. Create a new file there named `get-wallet-details.js` and define a function named `getWalletDetails` there. This function uses the [account_info method](account_info.html) to fetch account details and the [server_info method](server_info.html) to calculate the current [reserves](reserves.html). The code to do all this is as follows:
```

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/get-wallet-details.js", language="js") }}

> 5. Create another file in helpers folder named `get-ledger-details.js` which will be used to fetch the ledger details and copy the code written below to the file. In the function `getLedgerDetails`, we are using the [ledger](https://xrpl.org/ledger.html) that will fetch the details of the latest ledger.

Reviewer: ```suggestion
5. In the `helpers/` folder, create another file named `get-ledger-details.js`. Define a function, `getLedgerDetails()`, which fetches the latest validated ledger using the [ledger method](ledger.html). The code is as follows:
```

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/get-ledger-details.js", language="js") }}

> 6. Now, let's add the code to `index.js` file to fetch the account and ledger details and display them on the home page. Copy the code written below to the `index.js` file. Here we render the wallet and ledger details using the functions we implemented earlier in `get-wallet-details.js` and `get-ledger-details.js`. In order to make sure we have up to date data, we call the `getLedgerDetails` function every 4 seconds, as that's about how long it takes for a ledger to close on mainnet. 

Reviewer: ```suggestion
6. Now, modify the `index.js` file to use these helper functions and display the account and ledger details on the home page. Call the `getLedgerDetails` function every 4 seconds to refresh the ledger information, because that's roughly how long it takes for a ledger to close on Mainnet. 
```

Though I would prefer if you refactor the `getLedgerDetails` to use a subscription, in which case you'll need to rephrase this to reflect the change in design.

Author: This has been resolved, updated the method

{{ include_code("_code-samples/build-a-wallet/js/index.js", language="js") }}

> The one other thing we do here is add two buttons - one to send XRP and one to view the transaction history of our account. They won't work just yet - we'll have to add those implementations next.

Reviewer: ```suggestion
The one other thing we do here is add events to two buttons - one to send XRP and one to view transaction history. They won't work just yet—we'll implement them in the next steps.
```

It would be even better to break the HTML and `index.js` into multiple versions, one for each major step, so that you have a "self-contained" working version for that step, and then you can show only the new bits in the next step (similar to how the Python version of this tutorial does it).

Author: Although I agree that going block by block is more descriptive, I find it a bit of an overhead because the developer now has to put all block pieces together, which takes a bit more time and also adds more jargon which can be summarised in one line comment.

But I am still open to change the code a bit if you or @JST5000 or @khancode suggest.

Now the basic setup of your application is ready, you can run one of these commands

```bash
yarn dev
```
or 
```bash
npm run dev
```

> in your terminal which will start your app in development mode. Your terminal will display an URL using which you can verify the changes in real-time.

Reviewer: It's best not to have a code block mid-sentence. And as above, I recommend _not_ doing both yarn and npm examples for everything.

```suggestion
Now the application is ready to run. You can start it in dev mode using the following command:

\```bash
npm run dev
\```

Your terminal should output a URL which you can use to open your app in a browser. This dev site automatically updates to reflect any changes you make to the code.
```

(the backslashes are there because otherwise my "suggestion" gets broken into multiple code blocks. But you should remove them for the real thing)

Author: Noted, added.

- Show what's happening on the ledger
- And add a little logo for fun

To make that happen, we need to connect to the XRP Ledger and look up the account and the latest validated ledger.

5. In the `src/` directory, make a new folder named `helpers`. Create a new file there named `get-wallet-details.js` and define a function named `getWalletDetails` there. This function uses the [account_info method](account_info.html) to fetch account details and the [server_info method](server_info.html) to calculate the current [reserves](reserves.html). The code to do all this is as follows:

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/get-wallet-details.js", language="js") }}

6. Now, let's add the code to `index.js` file to fetch the account and ledger details and display them on the home page. Copy the code written below to the `index.js` file. Here we render the wallet details using the function we defined in `get-wallet-details.js`. In order to make sure we have up to date ledger data, we are using the [ledger stream](subscribe.html#ledger-stream) to listen for ledger close events.

{{ include_code("_code-samples/build-a-wallet/js/index.js", language="js") }}

> Also, add the following two files, one in the `helpers` folder [render-xrpl-logo.js]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/src/helpers/render-xrpl-logo.js) and for the other one create a new folder named `assets` in `src` directory and add [xrpl.svg]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/src/assets/xrpl.svg) file. These files are used to render the XRPL logo for aesthetic purposes.

Reviewer 2: Please split these into two steps with bullets (otherwise it looks like explanatory text instead of required steps, and there are really two steps here)

```suggestion
7. In the `helpers` folder, add [render-xrpl-logo.js]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/src/helpers/render-xrpl-logo.js) to handle displaying a logo.

8. Finally create a new folder named `assets` in `src` directory and add [xrpl.svg]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/src/assets/xrpl.svg) file. 

These files are used to render the XRPL logo for aesthetic purposes.
```

> The one other thing we do here is add events to two buttons - one to send XRP and one to view transaction history. They won't work just yet—we'll implement them in the next steps.

Reviewer 2: nit: 
```suggestion
The one other thing we do here is add events to two buttons - one to send XRP and one to view transaction history. They won't work just yet — we'll implement them in the next steps.
```

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/get-wallet-details.js", language="js") }}

6. Now, let's add the code to `index.js` file to fetch the account and ledger details and display them on the home page. Copy the code written below to the `index.js` file. Here we render the wallet details using the function we defined in `get-wallet-details.js`. In order to make sure we have up to date ledger data, we are using the [ledger stream](subscribe.html#ledger-stream) to listen for ledger close events.

{{ include_code("_code-samples/build-a-wallet/js/index.js", language="js") }}

7. In the `helpers` folder, add [render-xrpl-logo.js]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/src/helpers/render-xrpl-logo.js) to handle displaying a logo.

> 8. Finally create a new folder named `assets` in `src` directory and add [xrpl.svg]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/src/assets/xrpl.svg) file. 

Reviewer: ```suggestion
8. Finally create a new folder named `assets` in the `src/` directory and add the file [`xrpl.svg`]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/src/assets/xrpl.svg) there. 
```

You can use this code as a starting point for displaying your account's transaction history. If you want an additional challenge, try expanding it to support different transaction types (e.g. [TrustSet](trustset.html)). If you want inspiration for how to handle this, you can check out the [XRP Ledger Explorer](https://livenet.xrpl.org/) to see how the transaction details are displayed.

## Next Steps

Now that you have a functional wallet, you can take it in several new directions. The following are a few ideas:

- You could support more of the XRP Ledger's [transaction types](transaction-types.html) including [tokens](issued-currencies.html) and [cross-currency payments](cross-currency-payments.html)
- You could add support for displaying multiple tokens, beyond just XRP
- You could support creating [offers](offers.html) in the [decentralized exchange](decentralized-exchange.html)
- You could add new ways to request payments, such as with QR codes or URIs that open in your wallet.
- Or you could support better account security including allowing users to set [regular key pairs](cryptographic-keys.html#regular-key-pair) or handle [multi-signing](multi-signing.html).

> 

Reviewer: It could be good to have a bullet point directing people to the docs for [Building for Production](https://vitejs.dev/guide/build.html#public-base-path) with Vite, too.

Author: Added
